### DISCUSSION TITLE
Operand Error in Custom BC computeQpResidual

### DISCUSSION THREAD

**Discussion by smpeyres** [idaholab/moose#24576] [Q&A General] [answered]
I am trying to write a custom boundary condition for [Zapdos](https://github.com/shannon-lab/zapdos) that allows a user-prescribed flux in SI or similar units and is acceptable with the log-molar format used by the application.

From my work trying to move from strong to weak form, here's what I came up with mathematically (may not show up well in light mode):

![image](https://github.com/idaholab/moose/assets/107009840/bd76682b-64dc-49c8-8e30-383cbf206b95)

I tried to implement this in a .C and .h file, respectively. However, when I try to compile, it doesn't seem to like my usuage of the - sign between the diffusion term and the flux term:
```
/home/smpeyres_linux/projects/sean-zapdos/src/bcs/LogDiffusionFluxBC.C: In member function 'virtual ADReal LogDiffusionFluxBC::computeQpResidual()':
/home/smpeyres_linux/projects/sean-zapdos/src/bcs/LogDiffusionFluxBC.C:47:50: error: no match for 'operator-' (operand types are 'libMesh::CompareTypes<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> >::supertype' {aka 'MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>'} and 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> > >::type' {aka 'libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> >'})

   47 |          6.022e23 * _normals[_qp] * _grad_u[_qp] - _normals[_qp] * _value;
      |                                         `
```
I attached the .C, .h, and full error message in single zipped file.

[OperandError_BC_Zapdos.zip](https://github.com/idaholab/moose/files/11635619/OperandError_BC_Zapdos.zip)

Any advice or suggestions? Thank you!

Tagging the Zapdos experts: @cticenhour and @lindsayad

**Comment by GiudGiud** [comment]
Hello

```
for 'operator-' (
````
```
operand types are 'libMesh::CompareTypes<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> >::supertype' {aka 'MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true>'} and 
```
```
'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> > >::type' {aka 'libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<53> >, true> >'})
```

This means you are trying to substract a scalar (first one) from a vector
_normals[_qp] * _grad_u[_qp] is a scalar because of the dot product
 _normals[_qp] * _value is a vector as the product of a vector times a scalar

what are you really trying to compute?

Guillaume

**Comment by lindsayad** [comment]
Since Gamma is a vector it seems like what you want is for `_value` to be a vector

**Reply by smpeyres** [reply]
That's what I am thinking too. From the MOOSE framework I see "ADVectorFunctionDirichletBC" and "ADVectorFunctionNeumannBC". Should I try implementing something like this, and have a "flux" function prescribed later in the input file? Or is there something potentially simpler?

**Reply by lindsayad** [reply]
Can you just directly use `ADVectorFunctionNeumannBC`?

**Reply by smpeyres** [reply]
I would, but I am not confident that it would be compatible with the log densities, which have both the scalar value and gradient of the density in the weak form...

**Reply by lindsayad** [reply]
then it sounds like you should write your own object

**Reply by smpeyres** [reply]
Should I write it with  "ADVectorFunctionDirichletBC" and "ADVectorFunctionNeumannBC" as templates, or is there another BC in some module that would be more fitting?

**Reply by lindsayad** [reply]
Seems like Neumann is a pretty good fit as you want to prescribe a flux

**Reply by smpeyres** [reply]
[Error2_LogDiffusionFluxBC.zip](https://github.com/idaholab/moose/files/11735672/Error2_LogDiffusionFluxBC.zip)

I attempted to follow what the Neumann BC was doing, but I got two(?) errors. The first one, regarding '{' might be a consequence of the other error; I don't see anything immediately wrong with my brackets. The other error is again pointing to my computeQpResidual, and doesn't seem to like my exponential function. Any advice? Thank you!

**Reply by GiudGiud** [reply]
has a quick look:
```
    _function_z(getFunction("function_z")),
{ 
```
needs to become
```
    _function_z(getFunction("function_z"))
{ 
```

**Reply by GiudGiud** [reply]
I m not so sure about your computeQpResidual either, shouldnt the function be multiplied by the test function?

**Reply by lindsayad** [reply]
Yea your second term is a vector and needs to wind up being a scalar, e.g. just as @GiudGiud suggested you need to multiply by the test function.

I am a little concerned about the boundary condition you're imposing. When integrating by parts and looking at the result, it seems like you are imposing for your boundary condition:

flux = some-implicit-diffusive-flux + user-provided-flux

**Reply by smpeyres** [reply]
Removing the last comma after the getFunction fixed the bracket issue, but adding a _test[_i][_qp] term to the last part didn't fix the second error.

My philosophy for the boundary condition is that the first term in the computeQpResidual is the term that comes out from the weak form with the surface integral, which looks like the diffusive flux across the boundary. I decided that, like in a Dirichlet BC, the difference between the diffusive flux calculated based on D and N and the applied flux should approach zero. Is this the wrong way of looking at it?

I have a some of my work through the weak form in a PowerPoint deck, if y'all wish to see it.

**Reply by lindsayad** [reply]
No unfortunately that's not the correct way of looking at it. The sum of the residuals from all the integrated boundary conditions acting on a given boundary should be equal to what you want the total (outward) flux to be on that boundary to be.

**Reply by smpeyres** [reply]
Okay, I think I'm starting to understand... since the physics I am trying to describe is diffusion, then part of the weak form that pops out as the boundary condition is the surface integral of the flux. Then, and correct me if I am wrong with any of this:

1. If I do not provide a boundary condition, then it assumes that the flux is zero.
2. This why the Zapdos "ElectronDiffusionDoNothingBC.C" does nothing: you're just saying that the flux is equal to its mathematical definition, which is already the case, so it doesn't do anything.
3. Then, I simply need to tell the integral which value to integrate over, i.e., the user-provided flux. 

Seems like a simple VectorNeumannBC looks appropriate

**Reply by lindsayad** [reply, marked as answer]
Yes I think you've got it!

**Reply by smpeyres** [reply]
Awesome, thanks! So far looking pretty good, need to iron out some of the numerical kinks... See [https://github.com/shannon-lab/zapdos/discussions/200](https://github.com/shannon-lab/zapdos/discussions/200 )

**Reply by lindsayad** [reply]
Yea I saw that discussion. It looks pretty in-depth, so I may not dive into that one since it's out of my paid purview these days
